execution_engine/src/shared/gas_counter.rs

```diff
@@ -0,0 +1,113 @@
+pub mod error;
+mod large;
+mod small;
+
+use large::{LargeGasCounter, U64_MAX_AS_U512};
+use small::SmallGasCounter;
+
+use self::error::GasLimitError;
+
+use super::gas::Gas;
+
+/// Implements a gas counter functionality which uses specific implementation dependent on the size
+/// of gas limit to speed things up.
+#[derive(Debug, Copy, Clone)]
+pub enum GasCounter {
+    Small(SmallGasCounter),
+    Large(LargeGasCounter),
+}
+
+impl GasCounter {
+    pub fn new(limit: Gas, initial_count: Gas) -> GasCounter {
+        if limit.value() <= U64_MAX_AS_U512 && initial_count.value() <= U64_MAX_AS_U512 {
+            // Small counter if values fit in `0..u64::MAX`
+            GasCounter::Small(SmallGasCounter::new(
+                limit.value().as_u64(),
+                initial_count.value().as_u64(),
+            ))
+        } else {
+            GasCounter::Large(LargeGasCounter::new(limit, initial_count))
+        }
+    }
+
+    pub fn add(&mut self, additional_gas: u64) -> Result<(), GasLimitError> {
+        match self {
+            GasCounter::Small(counter) => counter.add(additional_gas),
+            GasCounter::Large(counter) => counter.add(additional_gas),
+        }
+    }
+
+    pub fn add_large(&mut self, additional_gas: Gas) -> Result<(), GasLimitError> {
```

Comment:
I think it'd be worth documenting why we have two separate methods for adding gas to discourage someone in the future just merging the two `add` methods.